### DISCUSSION TITLE
Why clangd didn't use lsp specified json format for configuration?

### DISCUSSION THREAD

**Discussion by nothankyouzzz** [clangd/clangd#1323] [Ideas]
I'm a neovim user, and tried use built-in lsp to config clangd, but I found I must ues non-json format to config it.

It's a lsp, but do not obey the lsp specification?

**Comment by sam-mccall** [comment]
Can you be more specific about which LSP feature you're talking about for configuration?

There are several we evaluated and found unsuitable, but possible we missed some.
Requirements include:
 - somewhat human-readable, in particular supporting comments
 - configuration can reflect user preferences (user-local) or project needs (checked in)
 - different configuration can be provided e.g. based on file extension or subdirectory
 - query config for each file in the project efficiently (for loading the index on startup etc), this precludes a per-file RPC to the editor such as workspace/configuration
 - mechanism works with all clients (AFAIK not all LSP clients easily expose e.g. initializeparams)
 - recommended settings can be documented in a way that's editor-independent and easy to follow (e.g. copy/paste snippets without translating escaping/quoting between syntaxes)

**Reply by nothankyouzzz** [reply]
The [LSP Specification](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_configuration) says: users could use a request (`workspace/configuration`) from server to config LSP Server. In neovim, particular to `lspconfig`, we cloud config lsp server like this:

```lua
lspconfig.clangd.setup {
  settings = {
    -- lsp configuration
  }
}
```

but clangd website only says user could config their lsp via `config.yaml`...

**Reply by sam-mccall** [reply]
Right. `workspace/configuration` wasn't suitable for the type of configuration we wanted to provide, because:
 - it performs too poorly (e.g. on startup, querying the configuration for 10k+ files can be on the critical path to loading the index.)
 - it doesn't have consistent support across editors

**Reply by nothankyouzzz** [reply]
> Right. `workspace/configuration` wasn't suitable for the type of configuration we wanted to provide, because:
> 
> * it performs too poorly (e.g. on startup, querying the configuration for 10k+ files can be on the critical path to loading the index.)
> * it doesn't have consistent support across editors

Q:

1. Just like `config.yaml`, due to it's a global configuration, so lsp server need only to query client (editor) for once? (i can't understand what *querying the configuration for 10k+ files* means, need i use 10k config files, but i just need some configuration for linter checks here.)
2. Use external file could be less consistent because user originally could use built-in lsp to config server instead a `config.yaml`. For  vscode, it should provide a lsp interface instead of a variety of lsp server plugins... (personal opinions)

**Reply by sam-mccall** [reply]
> i can't understand what querying the configuration for 10k+ files means

Configuration can differ for each file, e.g. based on the file's extension or path. (Requirement 3 above). `workspace/configuration` doesn't provide any mechanism for discovering this other than querying every file individually. Therefore we need 10k RPCs whether you actually have 10k config files or not.
We do effectively evaluate config.yaml (and any other config files that apply) 10k times in this scenario with lots of caching. This is fast enough, but the LSP version would not be.

> Use external file could be less consistent because user originally could use built-in lsp to config server instead a config.yaml.

If there was a question here, I missed it.
